### ISSUE TITLE
test-all-scream -- "dirty" git branch

### ISSUE THREAD

**Issue by AaronDonahue** [E3SM-Project/scream#214]
Would it be possible to adjust `test-all-scream` to allow for "dirty" branch testing when the `-b HEAD` option is exercised?  

I mean dirty branch, as a *not clean* branch, that hasn't had recent changes committed.  

The reason I ask is that I often use `test-all-scream` to make sure my changes won't break any of the unit tests, often I find an issue and have to add one (or sometimes more) commit to fix a small bug that impacts the unit tests.  It would be great for the commit chain if I didn't have to commit code I wasn't 100% sure of, then test, fix and commit again.

I'm guessing the requirement of a "clean" branch is that `test-all-scream` will checkout master at somepoint to generate baselines.  But if `-b HEAD` is used there is no need to do a git checkout and the need for a clean branch goes away (I suspect).  but I may be missing something.

**Comment by bartgol** [comment]
Uhm, `test-all-scream` should invoke `git checkout blah` only if the baseline ref is not `HEAD`. Can you show the command you launch (with all options) as well as the error msg?

**Comment by AaronDonahue** [comment]
```[donahue5@quartz1916:scream]$ git status
On branch aarondonahue/p3_robust_warnings
Changes not staged for commit:
  (use "git add <file>..." to update what will be committed)
  (use "git checkout -- <file>..." to discard changes in working directory)

	modified:   CMakeLists.txt

no changes added to commit (use "git add" and/or "git commit -a")
[donahue5@quartz1916:scream]$ git diff
diff --git a/components/scream/CMakeLists.txt b/components/scream/CMakeLists.txt
index 1bac4fa..4b2fb51 100644
--- a/components/scream/CMakeLists.txt
+++ b/components/scream/CMakeLists.txt
@@ -1,3 +1,4 @@
+# I'm adding a comment to dirty the branch
 if (NOT DEFINED PROJECT_NAME)
   cmake_minimum_required(VERSION 3.3)
   cmake_policy(SET CMP0057 NEW)
[donahue5@quartz1916:scream]$ ./scripts/test-all-scream $(which mpicxx) -m quartz -t sp -b HEAD
ERROR: Repo must be clean before running```

**Comment by AaronDonahue** [comment]
I think it's just a simple check.  so it doesn't get to the git checkout part, it just checks that the branch is clean.

**Comment by bartgol** [comment]
Ah, the issue is in `test-all-scream`, while I was looking in `test_all_scream.py`. This should be a quick fix.

**Comment by AaronDonahue** [comment]
I thought as much, I just wanted to make sure there wasn't another reason for a "clean branch".  Thanks for jumping on this @bartgol  :-)

**Comment by bartgol** [comment]
Uhm, I was about to issue the PR, but then I thought of a possible downside. If the pros outweigh the cons, then I'll go ahead and issue the PR, but first I want to bring this up.

The dirty-branch run can cause issues if one then forgets to commit the changes before pushing the branch, since the repo will now have an untested branch. It is true that the AT should catch those errors, so it is probably fine. And to be fair, forcing one to commit changes may result in placeholder commits with messages such as "try" or other non-informative stuff, which the user may later forget to rewrite before pushing the branch.

All in all, I'm for allowing dirty-branch testing, but I want to hear from @jgfouca as well.

**Comment by AaronDonahue** [comment]
You shouldn't be able to "push" a dirty branch right?

**Comment by bartgol** [comment]
Yes you can. You can always push whatever has been committed. Git push doesn't care what's the status of the working tree, it only cares about commits history, and whether you are ahead of the remote branch.

**Comment by bartgol** [comment]
Fixed by #215
